File: Data Structures & Algorithms/house-robber-ii/submission-0.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        if (n == 2) return max(nums[0], nums[1]);

        return max(solve(nums, 0, n - 2), solve(nums, 1, n - 1));
    }

private:
    int solve(vector<int>& nums, int start, int end) {
        int size = end - start + 1;
        vector<int> dp(size + 1, 0);
        
        dp[1] = nums[start];
        for (int i = 2; i <= size; i++) {
            // dp[i] corresponds to nums[start + i - 1]
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[start + i - 1]);
        }
        return dp[size];
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-0.py
import math

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # h = num of hours you have to eat all the bananas
        # piles[i] is the number of bananas in the ith pile
        # rate of eating: k (banana/ per hour)
        low, high = 1, max(piles)
        res = high

        while low <= high:
            mid = (low + high) // 2

            total = 0
            for pile in piles:
                total += math.ceil(pile / mid)
            
            if total <= h:
                res = mid
                high = mid - 1
            else:
                low = mid + 1
        return res
            


 




"""

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low, s = 0, len(nums)
        high = s - 1

        while low <= high:
            mid = (low + high) // 2
            if (nums[mid] == target):
                return mid
            elif nums[mid] < target:
                low = mid + 1
            else:
                high = mid - 1
        return -1
        

"""